NetKet/GroundState/variational_montecarlo.hpp

```diff
@@ -377,71 +501,7 @@ class VariationalMonteCarlo {
     auto pars = psi_.GetParameters();
 
     if (dosr_) {
-      const int nsamp = vsamp_.rows();
-
-      Eigen::VectorXcd b = Ok_.adjoint() * elocs_;
-      SumOnNodes(b);
-      b /= double(nsamp * totalnodes_);
-
-      if (!use_iterative_) {
-        // Explicit construction of the S matrix
-        Eigen::MatrixXcd S = Ok_.adjoint() * Ok_;
-        SumOnNodes(S);
-        S /= double(nsamp * totalnodes_);
-
-        // Adding diagonal shift
-        S += Eigen::MatrixXd::Identity(pars.size(), pars.size()) *
-             sr_diag_shift_;
-
-        Eigen::VectorXcd deltaP;
-        if (use_cholesky_ == false) {
-          Eigen::FullPivHouseholderQR<Eigen::MatrixXcd> qr(S.rows(), S.cols());
-          qr.setThreshold(1.0e-6);
-          qr.compute(S);
-          deltaP = qr.solve(b);
-        } else {
-          Eigen::LLT<Eigen::MatrixXcd> llt(S.rows());
-          llt.compute(S);
-          deltaP = llt.solve(b);
-        }
-        // Eigen::VectorXcd deltaP=S.jacobiSvd(ComputeThinU |
-        // ComputeThinV).solve(b);
-
-        assert(deltaP.size() == grad_.size());
-        grad_ = deltaP;
-
-        if (sr_rescale_shift_) {
-          Complex nor = (deltaP.dot(S * deltaP));
-          grad_ /= std::sqrt(nor.real());
-        }
-
-      } else {
-        Eigen::ConjugateGradient<MatrixReplacement, Eigen::Lower | Eigen::Upper,
-                                 Eigen::IdentityPreconditioner>
-            it_solver;
-        // Eigen::GMRES<MatrixReplacement, Eigen::IdentityPreconditioner>
-        // it_solver;
-        it_solver.setTolerance(1.0e-3);
-        MatrixReplacement S;
-        S.attachMatrix(Ok_);
-        S.setShift(sr_diag_shift_);
-        S.setScale(1. / double(nsamp * totalnodes_));
-
-        it_solver.compute(S);
-        auto deltaP = it_solver.solve(b);
-
-        grad_ = deltaP;
-        if (sr_rescale_shift_) {
-          auto nor = deltaP.dot(S * deltaP);
-          grad_ /= std::sqrt(nor.real());
-        }
-
-        // if(mynode_==0){
-        //   std::cerr<<it_solver.iterations()<<"
-        //   "<<it_solver.error()<<std::endl;
-        // }
-        MPI_Barrier(MPI_COMM_WORLD);
-      }
+      grad_ = sr_.ComputeUpdate(Ok_, 0.5 * grad_);
```

Comment:
Is it correct that the right-hand side passed to SR here is _half_ the gradient? This is what was used previously, because it was computed as
```cpp
Eigen::VectorXcd b = Ok_.adjoint() * elocs_;
SumOnNodes(b);
b /= double(nsamp * totalnodes_);
```
while the gradient is 
```cpp
 grad_ = 2. * (Ok_.adjoint() * elocs_);
// ...
```

Reply:
Yes, there is a factor of two, but it's not important since it provides a global rescaling to the learning rate if you want...